# Importing Libraries

In [54]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')
from google.colab import files
uploaded = files.upload()


Saving loan_data.csv to loan_data (2).csv


In [36]:
df = pd.read_csv('loan_data.csv')
df.head()

person_age person_gender person_education  person_income  person_emp_exp  \
0        22.0        female           Master        71948.0               0   
1        21.0        female      High School        12282.0               0   
2        25.0        female      High School        12438.0               3   
3        23.0        female         Bachelor        79753.0               0   
4        24.0          male           Master        66135.0               1   

  person_home_ownership  loan_amnt loan_intent  loan_int_rate  \
0                  RENT    35000.0    PERSONAL          16.02   
1                   OWN     1000.0   EDUCATION          11.14   
2              MORTGAGE     5500.0     MEDICAL          12.87   
3                  RENT    35000.0     MEDICAL          15.23   
4                  RENT    35000.0     MEDICAL          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  \
0                 0.49                         3.0           561   
1                 0.08                         2.0           504   
2                 0.44                         3.0           635   
3                 0.44                         2.0           675   
4                 0.53                         4.0           586   

  previous_loan_defaults_on_file  loan_status  
0                             No            1  
1                            Yes            0  
2                             No            1  
3                             No            1  
4                             No            1

# Shape of the Dataset

In [8]:
df.shape

(45000, 14)

# Unique Value

In [9]:
df['loan_status'].unique()

array([1, 0])

# Dataset

The Dataset given is a collection of datas of information about people who have taken loan . It firstly includes general information such as age , gender , education. Then it collects data regarding loan conditins such as income , home ownership ,loan intent . And finally all loan related information loan amount , interest rate , credit score etc.


# Preprocessing the Data

Removing irrevelant columns

In [37]:
columns = ['person_gender' ,'person_education' , 'person_home_ownership' , 'loan_intent' , 'previous_loan_defaults_on_file']
df = df.drop(columns, axis =1)
df.head()


person_age  person_income  person_emp_exp  loan_amnt  loan_int_rate  \
0        22.0        71948.0               0    35000.0          16.02   
1        21.0        12282.0               0     1000.0          11.14   
2        25.0        12438.0               3     5500.0          12.87   
3        23.0        79753.0               0    35000.0          15.23   
4        24.0        66135.0               1    35000.0          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  loan_status  
0                 0.49                         3.0           561            1  
1                 0.08                         2.0           504            0  
2                 0.44                         3.0           635            1  
3                 0.44                         2.0           675            1  
4                 0.53                         4.0           586            1

In [51]:
df = df.dropna()
df.head()

person_age  person_income  person_emp_exp  loan_amnt  loan_int_rate  \
0        22.0        71948.0               0    35000.0          16.02   
1        21.0        12282.0               0     1000.0          11.14   
2        25.0        12438.0               3     5500.0          12.87   
3        23.0        79753.0               0    35000.0          15.23   
4        24.0        66135.0               1    35000.0          14.27   

   loan_percent_income  cb_person_cred_hist_length  credit_score  loan_status  
0                 0.49                         3.0           561            1  
1                 0.08                         2.0           504            0  
2                 0.44                         3.0           635            1  
3                 0.44                         2.0           675            1  
4                 0.53                         4.0           586            1

# Scaling



In [62]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X = scaler.fit_transform(df.drop('loan_status' , axis =1))


# Encoding


In [64]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
y = encoder.fit_transform(df['loan_status'])

# Splitting Of Dataset

In [65]:
X_train , X_test , y_train , y_test = train_test_split(X,y,train_size=0.75,test_size=0.25,random_state=55)


# Imported Logistic Regression Model 1


In [66]:
M1 = LogisticRegression()
M1.fit(X_train,y_train)
y_pred=M1.predict(X_test)
M1accuracy = accuracy_score(y_test,y_pred)*100

# Model 2

In [70]:
class LRegression() :
  def __init__(self,learning_rate,iterations) :
    self.learning_rate = learning_rate
    self.iterations = iterations

  def fit( self , X , y) :
   self.m , self.n = X.shape
   self.W = np.zeros( self.n)
   self.b = 0
   self.X = X
   self.y = y

   for i in range( self.iterations ):
       self.update_weights()
       return self

  def update_weights( self ) :
    w = 1/(1+np.exp(-(self.X.dot(self.W)+self.b)))

    tmp = ( w - self.y.T)
    tmp = np.reshape( tmp , self.m)
    dW = np.dot( self.X.T , tmp)/self.m
    db = np.sum( tmp)/self.m

    self.W = self.W - self.learning_rate * dW
    self.b = self.b - self.learning_rate * db
    return self

  def predict( self , X) :
    Z = 1/(1 + np.exp(-(X.dot(self.W) + self.b)))
    y = np.where( Z > 0.5 , 1 , 0)
    return y











# Model Training



In [71]:
M2 = LRegression(learning_rate = 0.01 , iterations = 1000)

M2.fit( X_train, y_train)
M1 = LogisticRegression()
M1.fit(X_train,y_train)

y_pred = M2.predict( X_test )
y_pred1 = M1.predict( X_test)

correctly_classified = 0
correctly_classified1 = 0

for count in range( np.size( y_pred)) :

    if y_test[count] == y_pred[count] :
      correctly_classified = correctly_classified + 1

    if y_test[count] == y_pred1[count] :
      correctly_classified1 = correctly_classified1 + 1

      count = count + 1

print (" Accuracy of M2 : " , ( correctly_classified / count )*100)
print (" Accuracy of M1 : "  , ( correctly_classified1 / count)*100)

 Accuracy of M2 :  77.33131833940796
 Accuracy of M1 :  82.86958840785847


# Overall Analysis

The Model M2 made from sctrach using Logistic Regression gives an accuracy 77.331318.....96 while The Model M1 which was imported from sklearn gives an accuracy of 82.8695.........47. so The model provides better accuracy.